ClusterLoader design

wojtek-t opened the pull request:
@kubernetes/sig-scalability-feature-requests 
@sjug @countspongebob @shyamjvs

sjug commented on clusterloader/doc/vision.md:
````diff
@@ -0,0 +1,369 @@
+To define a test scenario, user will be providing the following config:
+
+```
+struct Config {
````
This is the biggest question mark, as the goal is to make running these tests accessible to users. While the data structures are robust in allowing CRUD operations as well as multiple steps, the configuration will be nested and elaborate.

The current configuration is much more declarative in nature, as we define:
```
ClusterLoader:
  projects:
    - num: 2
      basename: nginx-explorer
      tuning: default
      RCs:
        - num: 10
          file: nginx.yaml
        - num: 1
          file: explorer-pod.yaml
      pods:
        - num: 50
          image: k8s.gcr.io/pause-amd64:3.0
          basename: pausepods
          file: pod-pause.json
```
This is roughly equivalent to:
```
Config:
  Namespaces: 2
  Steps:
    Phases:
      Creates:	
      - NamespaceRange:
          min: 1
          max: 2
	ObjectType: RC
	ObjectTemplate: /path/nginx.yaml
	ReplicasPerNamespace: 10
	TuningSet: default
      - NamespaceRange:
          min: 1
          max: 2
	ObjectType: RC
	ObjectTemplate: /path/explorer-pod.yaml
	ReplicasPerNamespace: 1
	TuningSet: default
      - NamespaceRange:
          min: 1
          max: 2
	ObjectType: Pod
	ObjectTemplate: /path/pod-pause.json
	ReplicasPerNamespace: 50
	TuningSet: default
```

wojtek-t replied:
1. With the current definition, it's not possible to test Updates. So we need a concept of phase to achieve that (and that's pretty crucial).

2. Yes - it's roughly that (we may consider building-in the "namespace" deeper into that).
But after thinking much more about it in the meantime, I think that exposing explicitly some resources (e.g. pods, RCs etc. simply doesn't scale). and I want to be it generic so that all resources are treated roughly the same.

sjug replied:
It could also be possible to add states to the existing config type having a parent slice of projects, that way one can redefine the "state" at various points. Of course there would have to be additional logic to decide of create/update/delete between the states.

wojtek-t replied:
As discussed during SIG meeting yesterday - I agree that more declarative approach makes sense.
I refined the proposal a bit to accommodate that.

Also, I made it very clear what the logic for deciding create/update/delete should be to make it as simple as possible and understandable to users.

sjug commented on clusterloader/doc/vision.md:
````diff
@@ -0,0 +1,369 @@
+We define it as following:
+
+```
+struct PhaseDelete {
````
Is there some way to simply clean up all objects created?

wojtek-t replied:
Not now at least. For now the easiest way would be to:
- for every object type, create a phase delete selecting all pods (label.Everything()) from all namespace (by appropriate range)

I don't think it's very painful. WDYT?

sjug replied:
The problem would be if we have many object types the delete phase would be rather large.

wojtek-t replied:
We may consider making ObjectType to be slice of types.

Would you be ok with that?

wojtek-t replied:
After thinking about it a bit more (and changing the proposal a bit to make it declarative), I think this is no longer a problem.
1. Delete is pretty declarative now.
2. If you don't care about deletes as a test owner, you may simply don't care about it.
ClusterLoader itself at the end (as "AfterAllSteps" step) will be deleting all namespaces and that would automatically delete all objects from them too.

I made that explicit in the doc.

sjug commented on clusterloader/doc/vision.md:
```diff
@@ -0,0 +1,369 @@
+a referenced tuning set.
+
+
+### Measurements
```
I like building in the measurement capture into Cluster Loader..

wojtek-t replied:
Thanks!

sjug commented on clusterloader/doc/vision.md:
````diff
@@ -0,0 +1,369 @@
+
+```
+struct Step {
+	// Exactly one of these slices should be non-empty.
````
Why can a step only have a phase or a measurement?

wojtek-t replied:
What else do you suggest?

sjug replied:
Just not sure why they cant both happen in one step? I guess one would have to happen before the other if they were both defined?

wojtek-t replied:
Ordering was my main motivation. I would like to really make sure that we are able to start measurement before we start tests (or some of its part) and stop after everything is done.
I didn't want to introduce ordering within a phase to not overcomplicate things.

wojtek-t replied:
I made that explicit in the doc.

sjug commented on clusterloader/doc/vision.md:
````diff
@@ -0,0 +1,369 @@
+}
+```
+
+With a test being defined by a single json-file, it should be pretty simple to
````
Will yaml be compatible, as it is far more human readable?

wojtek-t replied:
Yes - we should make it work for both json and yaml.

sjug replied:
Sounds good

wojtek-t replied:
Done

wojtek-t commented on clusterloader/doc/vision.md:
```diff
@@ -0,0 +1,369 @@
+	RandomizedLoad *RandomizedLoad
+}
+
+// QpsLoad defines a uniform load with a given QPS.
```
Not sure I understand this comment.
QPS - is "queries per second" - I don't understand the second part.

wojtek-t commented on clusterloader/doc/vision.md:
````diff
@@ -0,0 +1,369 @@
+To make it usable, we introduce the following type:
+
+```
+struct Range {
````
In general, we need to make it possible to make some namespace different than others. So we need to make it possible to create something in just subset of namespaces.

sjug replied:
Is that for consistent addressing of the namespaces as well? namespace-min to namespace-max?

wojtek-t replied:
Yes - exactly.

wojtek-t commented on clusterloader/doc/vision.md:
````diff
@@ -0,0 +1,369 @@
+We define Tuning Set as following:
+
+```
+struct TuningSet {
````
I'm fine with making another tuning set (e.g. sth like:
```
type StepLoad struct {
  BurstSize int32
  Pause  time.Duration
}
```

Will that address your concern?

[In general, I think that we may need more in the future. I don't consider the stepping one as crucial initially, but if you really think it's needed, I'm fine with that.]

sjug replied:
Yeah that works, we use it very regularly.

jeremyeder replied:
It's a basic requirements. Allows us to drill down into efficiency metrics and prom data at particular load levels, often when we see a noticeable change  in relevant metrics

wojtek-t commented:
@krzysied @wasylkowski-a

shyamjvs commented on clusterloader/docs/vision.md:
````diff
@@ -0,0 +1,322 @@
+
+```
+struct Config {
+	// TBD: Way to access cluster.
````
I think the cannonical way we're doing this today in our testing is [passing path to](https://github.com/kubernetes/kubernetes/blob/master/hack/ginkgo-e2e.sh#L78) kubeconfig file. That's probably easier than passing address, cert, etc. Maybe let's use that?

wojtek-t replied:
I'm fine with that (it's a bit of an implementation detail though).

shyamjvs replied:
In that case, maybe we shouldn't be mentioning it here and then add a TBD? :)

wojtek-t replied:
It should be mentioned here, because it will be here. I added "path to kubeconfig" as possible solution here.

shyamjvs commented on clusterloader/docs/vision.md:
````diff
@@ -0,0 +1,322 @@
+```
+struct Step {
+	// Only one of Phases or Measurements can be non-empty.
+	Phases       []Phase
````
Is there a reason why you need to make these fields arrays? We're already having an array of Steps at the higher-level right? Unnecessary nesting can complicate our json.

wojtek-t replied:
yeah - good point. Changed to pointers.

shyamjvs commented on clusterloader/docs/vision.md:
````diff
@@ -0,0 +1,322 @@
+
+```
+struct Phase {
+	Declarations []PhaseDeclare
````
The same here.. You can simply have `Step = { *PhaseDeclare or *Measurement }` instead of one more struct and array.

wojtek-t replied:
That isn't true - this would reduce what can be expressed.
Having here []PhaseDeclare allows you to e.g. in parallel create one set of object, create the second set of objects and delete the third of objects.
So basically, PhaseDeclare`s within a single phase are being executed in parallel.

wojtek-t replied:
That said, we can probably get rid of `Phase` here and change Step to be either []PhaseDeclare or Measurement.

I will change that and make explicit in the doc.

shyamjvs replied:
> I will change that and make explicit in the doc.

Yes, thanks.

And wrt your parallelization argument, let me think a bit more if this is the way we want to formulate it.

wojtek-t replied:
Actually, I simplified it a bit. Now we only have a concept of "Step", "Phase" and "Measurement" and I got rid of "PhaseDeclare" at all.

PTAL

shyamjvs commented:
@wojtek-t This lgtm at a high-level. While I continue to review, a suggestion from my side -

It would be much easier to comprehend the doc if you structure it in a top-down manner IMO. Personally, it was a bit hard for me to keep track of each individual structs that you were introducing and then slowly trying to relate them. Instead if you started with defining what a Testconfig is (i.e a series of steps) and then what a step is (i.e a measurement or a phase) and so on, it'll be easier to understand.

shyamjvs commented:
> you can read it from the end to beginning :-)

Trust me.. This is what I ended up doing (to save time).

> understand the building blocks and then I understand what the final thing is

This might be really good approach when you're coming up with the design. But as someone (like me) who's reading the doc and isn't already familiar with your line of thought when you were writing it, it's harder to reason why you're doing it that way.. until when I reach the end and the suspense is finally revealed :)

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,320 @@
+
+// QpsLoad defines a uniform load with a given QPS.
+struct QpsLoad {
+	InitialDelay time.Duration
```
I sense that InitialDelay is a characteristic that can be common across the strategies.
Do you think we should move it up into TuningSet?

wojtek-t replied:
Yes - done.

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,320 @@
+
+// RandomizedLoad defines a load that is spread randomly
+// across a given total time.
+struct RandomizedLoad {
```
Can you help me understand how this is different from QpsLoad?

wojtek-t replied:
QPSLoad is that we generate exactly "QPS" queries per second (so one query every 1/QPS seconds).
RandomizedLoad is that we randomize the load over such period of time.

They are pretty similar in general.

[As an example:
- RandomizedLoad is how we spread over time things in load test
- QpsLoad is how we spread over time creations of latency pods]

shyamjvs replied:
I see.. thanks.

In that case, I believe it'll be much better to have `TimeSpread` instead of `AverageQps` (as you probably don't want to calculate it's value based on TimeSpread and #calls).

wojtek-t replied:
I'm not sure I understand it.
I think that average qps is exactly what I would like to pass.

IIUC, `TimeSpread` you're proposing correctly, would mean that I would like to create everything over a given time - this for me would be a separate policy (which we may want to add in the future, but for the purpose of migrating our existing e2e tests it won't be needed)

shyamjvs replied:
Ok.. let me instead ask this question. What do you think should be the averageQps for the scaling operation in our load test?

wojtek-t replied:
Currently we have exactly that - it's by default set to 10.

I don't want to use timespread, because then if you say change the size of the cluster - they you only need to change the number of objects and not change the tunning set.

wojtek-t replied:
> That's the expected throughput we want for pod churn. I believe what we want here is the rate at which we're going to change the RC objects as those are what would be involved in the Phase operations IIUC.

Yes - that was kind of shortcut from my side.
So let's assume that we want to have 10pods/s churn on average.
Then, if small RCs are 1/2 of all pods, then for these we want to generate 5pods/s. Given they have numReplicas=5, we would like to generate 1 per s on average.
if medium RCs are 1/4 of all pods, then we want to generate 2.5pods/s. Given they have numReplicas=25 (IIRC), that would mean an average load of 0.1 per s.
and so on

Yes - it requires some match to be done by someone creating a test. We may consider changing our tests in the future. But I think the steps should be:
- migrate to cluster loader
- then start changing tests if we want
We shouldn't do the two steps at the same time.

shyamjvs replied:
Makes sense.. thx for explaining.

shyamjvs commented on clusterloader/docs/vision.md:
````diff
@@ -0,0 +1,318 @@
+We define the phase as following:
+
+```
+struct Phase {
````
I can't think of a way in which we can use this way of defining Phase to for e.g doing sth like scale RCs from a given template to different values of `replicas` (like how we're doing in our load test).

So in the current form, IIUC this either means having 16k templates i.e one per RC (which I don't think we want to do) or scaling all RCs using a given template to the same no. of replicas (which can potentially change scenario of our load test).

shyamjvs replied:
> which can potentially change scenario of our load test

Maybe that's fine, but I'm not sure.

wojtek-t replied:
> I can't think of a way in which we can use this way of defining Phase to for e.g doing sth like scale RCs from a given template to different values of replicas (like how we're doing in our load test).

You can - the template will contain:
```
"numReplicas: "3+Rand()%5"
```

We don't need to change the scenario of our tests :)

shyamjvs replied:
That's a cool idea! One thing to note however is this might mean evaluating the template 16k times (which is probably fine).

wojtek-t replied:
Yes - that would require mutliple evaluations. I think it's fine, but it's a good point that it requires a note - adding.

wojtek-t replied:
Added to caveats section.

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,326 @@
+struct Step {
+	// Only one of these can be non-empty.
+	Phases       []Phase
+	Measurements []string
```
This is still the same. Aren't we moving it to single entry?

wojtek-t replied:
I reverted. The reasons are:
- I got rid of "DeclarativePhase": Phase is now effectively what previously was "DeclarativePhase".
 We need to have a slice of those, if we want to run couple different things in parallel.
- if we need a slice for phases, let's make it consistent for measurements :)

shyamjvs replied:
Ok, I think we can live with it.

shyamjvs commented on clusterloader/docs/vision.md:
````diff
@@ -0,0 +1,326 @@
+
+```
+struct Step {
+	Phases []Phase
````
This is a bit confusing. You've redefined the same struct later in the document differently (with Measurements).

wojtek-t replied:
Done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,326 @@
+struct Phase {
+	// Set of namespace in which objects should be reconciled.
+	// If null, objects are assumed to be cluster scoped.
+	NamespaceRange *Range
```
Sg.. thanks!

shyamjvs commented:
Btw - I still suggest having top-down approach. It's very easy (at least for me) to form a mental picture if you can begin with saying "test config is a list of steps" and then "a step is a set of measurements or phases done in parallel". In the current form, I'm having to read so many details about phases, measurements, etc even before reaching the foundational point of what overall test looks like.

shyamjvs commented:
We can probably verify this hypothesis with a different reader (as I might be biased) :)
@wasylkowski-a @krzysied

wojtek-t commented:
> It seems you feel extremely strongly about that.
> Since I don't have that strong opinion here, I will change that tomorrow.

This is now changed to be more top-down. PTAL if that's fine for you.

Also, I modified the `Phase` a bit to make it more flexible.

shyamjvs commented:
Thanks for making the change - will TAL soon.

> PTAL if that's fine for you.

Like we were discussing offline, not having definition of 'Config' at the top (i.e above the definition of 'Step') is making it a bit hard to follow. Maybe you could fix that?

wojtek-t commented:
> Like we were discussing offline, not having definition of 'Config' at the top (i.e above the definition of 'Step') is making it a bit hard to follow. Maybe you could fix that?

Done - PTAL

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,335 @@
+
+As of 31/03/2018, all our scalability tests, are regular e2e tests written in Go.
+It makes them really unfriendly for developers not working on scalability who just
+want to load tests Kubernetes features they are working on. Doing so in many
```
load test

wojtek-t replied:
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,335 @@
+As of 31/03/2018, all our scalability tests, are regular e2e tests written in Go.
+It makes them really unfriendly for developers not working on scalability who just
+want to load tests Kubernetes features they are working on. Doing so in many
+situation requires understanding how those tests really work, modify their code
```
situations

wojtek-t replied:
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,335 @@
+It makes them really unfriendly for developers not working on scalability who just
+want to load tests Kubernetes features they are working on. Doing so in many
+situation requires understanding how those tests really work, modify their code
+to test the new feature and only then run and debug tests (or create load test for
```
You could move this long sentence inside brackets to new sentence?

wojtek-t replied:
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,335 @@
+situation requires understanding how those tests really work, modify their code
+to test the new feature and only then run and debug tests (or create load test for
+your feature on your own, which may be easier, but on the other hand may not
+exercise important metrics that our performance test check). This workflow is far
```
tests

wojtek-t replied:
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,335 @@
+### Phase
+
+Phase declaratively defines state of objects we should reach in the underlying
+Kuberneces cluster. A single declaration may result in a number of create, update
```
Kubernetes

wojtek-t replied:
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,335 @@
+
+Phase declaratively defines state of objects we should reach in the underlying
+Kuberneces cluster. A single declaration may result in a number of create, update
+and delete operation depending on the current state of the cluster.
```
operations

wojtek-t replied:
done

wojtek-t commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,335 @@
+
+In early 2017 the prototype of “Cluster Loader” has been created. It proved that
+configuring tests with json/yaml files is possible. But its functionality is very
+limited and it was very far from enabling migration any of existing scalability
```
done

shyamjvs commented on clusterloader/docs/vision.md:
````diff
@@ -0,0 +1,335 @@
+
+```
+struct Phase {
+	// Set of namespace in which objects should be reconciled.
````
namespaces

wojtek-t replied:
done

wojtek-t commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,335 @@
+Additionally, we will introduce a set of predefined operations (that user will be
+able to use as phases). They will allow users to monitor/measure performance impact
+of user-defined phases.
+The following subsections will describe this in a bit more details.
```
Interesting - I didn't know that.

wojtek-t commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,335 @@
+
+At the high level, a single test will consist of a number of steps. If we have
+really good usecase, we may consider creating a DAG of steps, but for now we
+don't have any rationale for it. In each of steps we will be creating, updating
```
done

wojtek-t commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,335 @@
+to modify scenarios, and fork them to new ones.
+
+Note that before running any steps, ClusterLoader will create all the necessary
+namespace and after running all of them will delete them (together with all
```
done

wojtek-t commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,335 @@
+
+Note that before running any steps, ClusterLoader will create all the necessary
+namespace and after running all of them will delete them (together with all
+objects that remained undeleted after running the test). Namespace are described
```
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,334 @@
+	// Name of TuningSet to be used.
+	TuningSet string
+	// A set of objects that should be reconciled.
+	Objects []Object
```
I'm not sure I understand why this should be an array?
This would iiuc mean that we use same `ReplicasPerNamespace` to be applied for all the objects in the array (which I can't see much use of).

shyamjvs replied:
Also same applies wrt other configs like TuningSet and NamespaceRange.

shyamjvs replied:
Ok... I see you have an explanation below. Let me read it.

shyamjvs replied:
Read it and I still have some confusion. So IIUC:

- we already have steps that are serial
- within a step we have phases that are parallel
- with a phase we have objects that are again serially processed

What I don't understand is why it isn't enough to have just one serial construct and one parallel construct? Maybe we can discuss this f2f.

wojtek-t replied:
Discussed f2f and added rationale below.

shyamjvs commented:
So @wojtek-t and me discussed offline about this and agreed after discussing some use-cases that the current design is a reasonably good trade-off of usability and flexibility. I'm going to lgtm this (neglecting typos by will:) once explanations of few design choices are made more clearer. Also, we'd follow this up with a user-guide demonstrating how to use the tool for different use-cases.

wojtek-t commented:
> So @wojtek-t and me discussed offline about this and agreed after discussing some use-cases that the current design is a reasonably good trade-off of usability and flexibility. I'm going to lgtm this (neglecting typos by will:) once explanations of few design choices are made more clearer. Also, we'd follow this up with a user-guide demonstrating how to use the tool for different use-cases.

Added some explanations. PTAL if that is satisfying for you and if not let me know which part should be clarified.

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,344 @@
+## Vision
+
+At the high level, a single test will consist of a number of steps. If we have
+really good usecase, we may consider creating a DAG of steps, but for now we
```
Btw - I think this DAG argument can now be removed with the newer design. It's possible to achieve even a DAG of steps now with just a sequence of steps by merging nodes of the DAG by BFS level number. And by merging nodes, I mean joining their []Phases. Wdyt?

wojtek-t replied:
Yeah - I think it no longer makes much sense. Removed.

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,344 @@
+all actions are done.
+
+Also note that all `Phases` and `Measurements` within a single `Step` will be
+run in parallel - a `Step` ends when all its `Phases` or `Measurements` finish.
```
Are we also mentioning somewhere that the 'steps' themselves run serially?

wojtek-t replied:
Yes - it was part of rationale. But I also made that explicit here.

shyamjvs commented on clusterloader/docs/vision.md:
````diff
@@ -0,0 +1,344 @@
+```
+
+The semantic of the above structure will be as following:
+- `Phases` within a single `Step` will be run in parallel
````
Ok.. So here's a good place to clarify that the 'steps' are run serially (as I suggested in above comment).

wojtek-t replied:
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,344 @@
+- if `ReplicasPerNamespace` is the same as it previously was, we will update all
+  objects to the referenced template.
+
+An appropriate validation will be added to cluster loader to ensure that for a
```
s/that/the above/

wojtek-t replied:
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,344 @@
+An appropriate validation will be added to cluster loader to ensure that for a
+given input config.
+
+Note that (namespace number, object type, basename) tuple defines a set of object.
```
'a set of replicated objects' ?

wojtek-t replied:
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,344 @@
+
+The semantic of the above structure will be as following:
+- `Phases` within a single `Step` will be run in parallel
+- `Object` within a single `Phase` will be reconciled in serial for a given
```
Objects

wojtek-t replied:
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,344 @@
+The semantic of the above structure will be as following:
+- `Phases` within a single `Step` will be run in parallel
+- `Object` within a single `Phase` will be reconciled in serial for a given
+  (namespace, replica number pair). For different (namespace, replica number)
```
(namespace, replica number) pair

wojtek-t replied:
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,344 @@
+  pairs they will be spread using a given tuning set.
+
+The rationale for having such structure is the following:
+- Slice of `Object` allows for making a unit of work to be greater than
```
I'd suggest rewording this sentence as:

`Objects` represents a collection of k8s objects that can be logically thought of as a unit of workload (for e.g a web application that's comprised of a service, a deployment and a volume). Conceptually, this collection is our unit for replication. The replication itself is done according to the TuningSet and ReplicasPerNamespace parameters of the `Phase`. Also note that we process the `Objects` slice serially which allows ordering while processing objects of a unit (e.g create service before creating deployment).

wojtek-t replied:
Slightly modified - but thanks for making it clear.

shyamjvs commented on clusterloader/docs/vision.md:
````diff
@@ -0,0 +1,344 @@
+```
+
+Using this will allow us to easily use dynamic client in most of the places
+which may significantly simply Cluster Loader itself.
````
simplify

wojtek-t replied:
done

shyamjvs commented on clusterloader/docs/vision.md:
```diff
@@ -0,0 +1,344 @@
+	// of referenced namespaces.
+	ReplicasPerNamespace int32
+	// Name of TuningSet to be used.
+	TuningSet string
```
You are using name of tuningSet instead of the struct itself. Where are we defining those tuningSets then? Are they defined globally (I can't find them in `Config`)?

wojtek-t replied:
Good point - yes - they should be part of config. Added.
